File: app/src/main/java/com/artemchep/horario/_new/activities/SubjectActivity.java
package com.artemchep.horario._new.activities;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.ColorUtils;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.ViewGroup;

import com.artemchep.basic.Device;
import com.artemchep.basic.tests.Check;
import com.artemchep.horario.Config;
import com.artemchep.horario.R;
import com.artemchep.horario._new.fragments.FragmentSubject;
import com.artemchep.horario._new.fragments.SubjectAboutFragment;
import com.artemchep.horario._new.fragments.SubjectTasksFragment;
import com.artemchep.horario._new.fragments.SubjectMaterialsFragment;
import com.artemchep.horario._new.fragments.SubjectStreamFragment;
import com.artemchep.horario._new.fragments.SubjectStudentsFragment;
import com.artemchep.horario.database.models.SubjectInfo;
import com.artemchep.horario.ui.activities.ActivityHorario;
import com.artemchep.horario.ui.widgets.ContainersLayout;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;
import com.ogaclejapan.smarttablayout.SmartTabLayout;
import com.ogaclejapan.smarttablayout.utils.v4.FragmentPagerItemAdapter;
import com.ogaclejapan.smarttablayout.utils.v4.FragmentPagerItems;
import com.thebluealliance.spectrum.internal.ColorUtil;

import net.opacapp.multilinecollapsingtoolbar.CollapsingToolbarLayout;

/**
 * @author Artem Chepurnoy
 */
public class SubjectActivity extends ActivityHorario implements ViewPager.OnPageChangeListener {

    /**
     * @author Artem Chepurnoy
     */
    public interface Page {

        void setSubject(@Nullable SubjectInfo subject);

        void onFabClick(@NonNull View view);

        /**
         * @return {@code true} if this page should have
         * {@link android.support.design.widget.FloatingActionButton fab} shown,
         * {@code false} otherwise
         */
        boolean hasFab();
    }

    public static final String EXTRA_USER_ID = "extra::user::id";
    public static final String EXTRA_TIMETABLE_ID = "extra::timetable::id";
    public static final String EXTRA_SUBJECT = "extra::subject";

    private static final String TAG_FRAGMENT_DETAILS = "tag_details";
    private static final String TAG_FRAGMENT_MASTER = "tag_master";

    private String mUserId;
    private String mTimetableId;
    private SubjectInfo mSubject;

    private AppBarLayout mAppBar;
    private ContainersLayout mContainers;
    private SmartTabLayout mSmartTabLayout;
    private CollapsingToolbarLayout mCollapsingToolbar;
    private FragmentPagerItemAdapter mAdapter;
    private FloatingActionButton mFab;
    private ViewPager mViewPager;
    private View mBackdropToolbar;
    private Toolbar mToolbar;

    @NonNull
    private final ValueEventListener mEventListener = new ValueEventListener() {
        @Override
        public void onDataChange(DataSnapshot dataSnapshot) {
            if (dataSnapshot.exists()) {
                mSubject = dataSnapshot.getValue(SubjectInfo.class);
                mSubject.key = dataSnapshot.getKey();
                bind(mSubject);
            } //else supportFinishAfterTransition();
        }

        @Override
        public void onCancelled(DatabaseError databaseError) {
        }
    };

    @NonNull
    private final View.OnClickListener mFabOnClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            FragmentSubject fragment = (FragmentSubject) mAdapter.getPage(mViewPager.getCurrentItem());
            fragment.onFabClick(v);
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Apply current theme
        switch (Config.getInstance().getInt(Config.KEY_UI_THEME)) {
            case Config.THEME_LIGHT:
                setTheme(R.style.AppThemeLight_NoActionBar);
                break;
            case Config.THEME_DARK:
                setTheme(R.style.AppTheme_NoActionBar);
                break;
            case Config.THEME_BLACK:
                setTheme(R.style.AppThemeBlack_NoActionBar);
                break;
        }

        Bundle args = getIntent().getExtras();
        mUserId = args.getString(EXTRA_USER_ID);
        mTimetableId = args.getString(EXTRA_TIMETABLE_ID);
        mSubject = args.getParcelable(EXTRA_SUBJECT);
        // Make sure subject exists and
        // contains at least the key
        Check.getInstance().isNonNull(mSubject);
        Check.getInstance().isNonNull(mSubject.key);

        setContentView(R.layout.activity_subject);

        mFab = (FloatingActionButton) findViewById(R.id.fab);
        mFab.setOnClickListener(mFabOnClickListener);
        mAppBar = (AppBarLayout) findViewById(R.id.appbar);
//        mContainers = (ContainersLayout) findViewById(R.id.containers);
        mSmartTabLayout = (SmartTabLayout) mAppBar.findViewById(R.id.tabs);
        mCollapsingToolbar = (CollapsingToolbarLayout) mAppBar.findViewById(R.id.toolbar_collapsing);
        mBackdropToolbar = mCollapsingToolbar.findViewById(R.id.toolbar_backdrop);
        mViewPager = (ViewPager) findViewById(R.id.viewpager);
        mToolbar = (Toolbar) mCollapsingToolbar.findViewById(R.id.toolbar);
        mToolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                supportFinishAfterTransition();
            }
        });

        mAdapter = new FragmentPagerItemAdapter(
                getSupportFragmentManager(), FragmentPagerItems.with(this)
                .add(getString(R.string.__subject_stream), SubjectStreamFragment.class)
                .add(getString(R.string.__subject_tasks), SubjectTasksFragment.class)
                .add(getString(R.string.__subject_materials), SubjectMaterialsFragment.class)
                .add(getString(R.string.__subject_students), SubjectStudentsFragment.class)
                .add(getString(R.string.__subject_about), SubjectAboutFragment.class)
                .create()) {

            @Override
            public Object instantiateItem(ViewGroup container, int position) {
                Object item = super.instantiateItem(container, position);
                if (item instanceof Page) {
                    Page f = (Page) item;
                    f.setSubject(mSubject);
                }
                return item;
            }

        };

        mViewPager.setAdapter(mAdapter);
        mViewPager.addOnPageChangeListener(this);

        bind(mSubject);
    }

    @Override
    protected void onStart() {
        super.onStart();
        /*
        FirebaseDatabase.getInstance()
                .getReference(Dbb.SUBJECTS)
                .child(mSubject.key)
                .child(Dbb.SUBJECT_INFO)
                .addValueEventListener(mEventListener);
                */
    }

    @Override
    protected void onStop() {
        /*
        FirebaseDatabase.getInstance()
                .getReference(Dbb.SUBJECTS)
                .child(mSubject.key)
                .child(Dbb.SUBJECT_INFO)
                .removeEventListener(mEventListener);
                */
        super.onStop();
    }

    @SuppressLint("NewApi")
    private void bind(@NonNull SubjectInfo subject) {
        mCollapsingToolbar.setTitle(subject.name);

        int color = subject.color;
        color |= 0xFF000000; // ignore alpha bits
        final boolean isColorDark = ColorUtil.isColorDark(color);

        if (Device.hasLollipopApi()) {
            int statusBarColor = ColorUtils.blendARGB(color, Color.BLACK, 0.4f);
            getWindow().setStatusBarColor(statusBarColor);
        }

        Drawable overflowIcon;
        if (isColorDark) {
            mToolbar.setNavigationIcon(R.drawable.ic_keyboard_backspace_white_24dp);
            mCollapsingToolbar.setExpandedTitleColor(Color.WHITE);
            mCollapsingToolbar.setCollapsedTitleTextColor(Color.WHITE);
            overflowIcon = ContextCompat.getDrawable(this, R.drawable.ic_dots_vertical_white_24dp);
            if (mSmartTabLayout != null) {
                mSmartTabLayout.setDefaultTabTextColor(Color.WHITE);
                mSmartTabLayout.setSelectedIndicatorColors(Color.WHITE);
            }
        } else {
            mToolbar.setNavigationIcon(R.drawable.ic_keyboard_backspace_black_24dp);
            mCollapsingToolbar.setExpandedTitleColor(Color.BLACK);
            mCollapsingToolbar.setCollapsedTitleTextColor(Color.BLACK);
            overflowIcon = ContextCompat.getDrawable(this, R.drawable.ic_dots_vertical_black_24dp);
            if (mSmartTabLayout != null) {
                mSmartTabLayout.setDefaultTabTextColor(Color.BLACK);
                mSmartTabLayout.setSelectedIndicatorColors(Color.BLACK);
            }
        }

        mCollapsingToolbar.setContentScrimColor(color);
        if (mBackdropToolbar != null) {
            mBackdropToolbar.setBackgroundColor(color);
        } else mAppBar.setBackgroundColor(color);
        mToolbar.setOverflowIcon(overflowIcon);

        mSmartTabLayout.setViewPager(mViewPager);
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
    }

    @Override
    public void onPageSelected(int position) {
        Page page = (Page) mAdapter.getPage(position);
        if (page.hasFab()) {
            mFab.show();
        } else mFab.hide();
    }

    @Override
    public void onPageScrollStateChanged(int state) {
    }

    public void navigateDetailsFrame(@NonNull Fragment fragment) {
/*
        if (mAppBar.getState() == MainActivity.State.SINGLE_COLUMN_MASTER) {
            mAppBar.setState(MainActivity.State.SINGLE_COLUMN_DETAILS);
            mContainers.setState(MainActivity.State.SINGLE_COLUMN_DETAILS);
        } else {
            mAppBar.setState(MainActivity.State.TWO_COLUMNS_WITH_DETAILS);
            mContainers.setState(MainActivity.State.TWO_COLUMNS_WITH_DETAILS);
        }
        */

        findViewById(R.id.activity_main__frame_details).setVisibility(View.VISIBLE);
        getSupportFragmentManager().beginTransaction()
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                .replace(R.id.activity_main__frame_details, fragment, TAG_FRAGMENT_DETAILS)
                .commit();
    }

}

File: app/src/main/java/com/artemchep/horario/_new/activities/SubjectActivity.kt
package com.artemchep.horario._new.activities

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.support.annotation.DrawableRes
import android.support.design.widget.AppBarLayout
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.Fragment
import android.support.v4.app.NavUtils
import android.support.v4.graphics.ColorUtils
import android.support.v4.view.ViewPager
import android.support.v7.widget.Toolbar
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import com.afollestad.materialdialogs.DialogAction
import com.afollestad.materialdialogs.MaterialDialog
import com.artemchep.horario.Device
import com.artemchep.horario.Config
import com.artemchep.horario.Palette
import com.artemchep.horario.R
import com.artemchep.horario._new.UiHelper
import com.artemchep.horario.database.data.Role
import com.artemchep.horario.database.data.Subject
import com.artemchep.horario.database.models.ISubject
import com.artemchep.horario.extensions.withAll
import com.artemchep.horario.interfaces.FragmentHost
import com.artemchep.horario.ui.fragments.*
import com.artemchep.horario.ui.utils.EXTRA_COLOR
import com.artemchep.horario.ui.utils.EXTRA_SUBJECT
import com.artemchep.horario.ui.utils.EXTRA_SUBJECT_ID
import com.artemchep.horario.ui.utils.EXTRA_USER_ID
import com.google.firebase.firestore.FirebaseFirestore
import com.ogaclejapan.smarttablayout.SmartTabLayout
import com.ogaclejapan.smarttablayout.utils.v4.FragmentPagerItemAdapter
import com.ogaclejapan.smarttablayout.utils.v4.FragmentPagerItems
import com.thebluealliance.spectrum.internal.ColorUtil
import net.opacapp.multilinecollapsingtoolbar.CollapsingToolbarLayout

/**
 * @author Artem Chepurnoy
 */
class SubjectActivity : MultiPaneActivity(),
        ViewPager.OnPageChangeListener,
        Toolbar.OnMenuItemClickListener,
        View.OnClickListener {

    private lateinit var firestore: FirebaseFirestore
    private lateinit var userId: String
    private lateinit var subjectId: String

    private lateinit var memberDoc: FragmentDocument.FireDocument<Role>
    private lateinit var subjectUser: FragmentDocument.FireDocument<ISubject>
    private val manager = FragmentDocument.FireManager()

    private val adapterArgs = Bundle()

    private var pagePosition = -1 // invalid page as initial value

    private lateinit var fab: FloatingActionButton
    private lateinit var appBar: AppBarLayout
    private lateinit var tabLayout: SmartTabLayout
    private lateinit var toolbar: Toolbar
    private lateinit var toolbarCollapsing: CollapsingToolbarLayout
    private lateinit var refreshMenuItem: MenuItem
    private lateinit var moderateMenuItem: MenuItem
    private lateinit var viewPager: ViewPager
    private lateinit var adapter: FragmentPagerItemAdapter

    /**
     * @author Artem Chepurnoy
     */
    interface Page {

        fun onFabClick(view: View) {}

        /**
         * @return `true` if this page should have
         * [fab][android.support.design.widget.FloatingActionButton] shown,
         * `false` otherwise
         */
        @DrawableRes
        fun hasFab(): Int = 0
    }

    override var primaryColor: Int? = Palette.GREY
        get() = color

    /**
     * Interface color setter; Setting this actually changes the
     * accent color of the user interface.
     */
    private var color: Int = 0
        set(color) {
            if (field != color) {
                appBar.setBackgroundColor(color)
                toolbarCollapsing.setContentScrimColor(color)
                UiHelper.updateToolbarBackIcon(toolbar, color)
                UiHelper.updateToolbarMenuIcons(toolbar, color)
                UiHelper.updateCollapsingToolbarTitle(toolbarCollapsing, color)

                if (Device.hasLollipopApi()) {
                    window.statusBarColor = ColorUtils.blendARGB(color, Color.BLACK, 0.6f)
                }

                val isColorDark = ColorUtil.isColorDark(color)
                val tabColor = if (isColorDark) Color.WHITE else Color.BLACK
                tabLayout.apply {
                    setDefaultTabTextColor(tabColor)
                    setSelectedIndicatorColors(tabColor)
                    setViewPager(viewPager)
                }

                adapterArgs.putInt(EXTRA_COLOR, color)
            }

            field = color
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Apply current theme
        when (Config.get<Int>(Config.KEY_UI_THEME)) {
            Config.THEME_LIGHT -> setTheme(R.style.AppThemeLight_NoActionBar)
            Config.THEME_DARK -> setTheme(R.style.AppTheme_NoActionBar)
            Config.THEME_BLACK -> setTheme(R.style.AppThemeBlack_NoActionBar)
        }

        intent.extras!!.run {
            userId = getString(EXTRA_USER_ID)
            subjectId = getString(EXTRA_SUBJECT_ID)
        }

        firestore = FirebaseFirestore.getInstance()
        setupDatabase(savedInstanceState)

        setContentView(R.layout.___activity_subject)

        fab = findViewById<FloatingActionButton>(R.id.fab).apply {
            setOnClickListener(this@SubjectActivity)
        }
        appBar = findViewById(R.id.appbar)
        tabLayout = appBar.findViewById(R.id.tabs)
        toolbarCollapsing = appBar.findViewById(R.id.toolbar_collapsing)
        toolbar = toolbarCollapsing.findViewById<Toolbar>(R.id.toolbar).apply {
            setNavigationOnClickListener {
                val activity = this@SubjectActivity
                val intent = Intent(activity, MainActivity::class.java)
                intent.addCategory(MainActivity.CATEGORY_SUBJECTS)
                NavUtils.navigateUpTo(activity, intent)
            }
            // Setup menu
            inflateMenu(R.menu.___subject)
            setOnMenuItemClickListener(this@SubjectActivity)
            refreshMenuItem = menu.findItem(R.id.action_refresh)!!
            moderateMenuItem = menu.findItem(R.id.action_moderate)!!
        }

        adapterArgs.putString(EXTRA_USER_ID, userId)
        adapterArgs.putString(EXTRA_SUBJECT_ID, subjectId)
        adapter = object : FragmentPagerItemAdapter(
                supportFragmentManager, FragmentPagerItems.with(this)
                .add(R.string.__subject_stream, SubjectStreamFragment::class.java)
                .add(R.string.__subject_tasks, SubjectTasksFragment::class.java)
                .add(R.string.__subject_schedule, SubjectSchedulesFragment::class.java)
                .add(R.string.__subject_students, SubjectMembersFragment::class.java)
                .add(R.string.__subject_about, SubjectAboutFragment::class.java)
                .create()) {

            override fun instantiateItem(container: ViewGroup, position: Int): Any {
                val item = super.instantiateItem(container, position)
                if (item is Fragment) try {
                    item.arguments = adapterArgs
                } catch (ignored: IllegalStateException) {
                }
                return item
            }

            override fun setPrimaryItem(container: ViewGroup, position: Int, `object`: Any) {
                super.setPrimaryItem(container, position, `object`)
                onPageSelected(position)
            }

        }

        viewPager = findViewById<ViewPager>(R.id.viewpager).apply {
            adapter = this@SubjectActivity.adapter
            addOnPageChangeListener(this@SubjectActivity)
        }

        setupObservers()
    }

    private fun setupDatabase(savedInstanceState: Bundle?) {
        memberDoc = FragmentDocument.FireDocument(
                "subjects/$subjectId/members/$userId",
                "member", {
            Role().apply { inflateRole(it) }
        }).also(manager::link)

        val extras = intent.extras!!
        subjectUser = FragmentDocument.FireDocument<ISubject>(
                "users/$userId/subjects/$subjectId",
                "subject_user", { snapshot ->
            Subject(id = snapshot.id).apply { inflateSubject(snapshot) }
        }).also {
            it.value = extras.getParcelable(EXTRA_SUBJECT)
            manager.link(it)
        }

        // Restore data
        savedInstanceState?.run {
            memberDoc.value = getParcelable(memberDoc.tag)
            subjectUser.value = getParcelable(subjectUser.tag)
        }
    }

    private fun setupObservers() {
        manager.state.follow { bind() }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.fab -> {
                // Redirect click event to current
                // page fragment
                val pos = viewPager.currentItem
                val page = adapter.getPage(pos) as Page
                page.onFabClick(v)
            }
        }
    }

    override fun onMenuItemClick(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_refresh -> manager.refresh()
            R.id.action_moderate -> {
                val args = Bundle().apply {
                    putString(EXTRA_USER_ID, userId)
                    putString(EXTRA_SUBJECT_ID, subjectId)
                }

                // Start subject moderate activity
                val host = activity as FragmentHost
                host.fragmentShow(SubjectModerateFragment::class.java, args, FragmentHost.FLAG_AS_DIALOG)
            }
            R.id.action_configure -> {
                val args = Bundle().apply {
                    putString(EXTRA_USER_ID, userId)
                    putString(EXTRA_SUBJECT_ID, subjectId)
                }

                // Start subject configure activity
                val host = activity as FragmentHost
                host.fragmentShow(SubjectConfigureFragment::class.java, args, FragmentHost.FLAG_AS_DIALOG)
            }
            R.id.action_leave -> {
                val md = MaterialDialog.Builder(this)
                        .content("Leave subject? This will not remove your contributions")
                        .positiveText(R.string.action_leave)
                        .negativeText(android.R.string.cancel)
                        .onPositive { _, _ ->
                            firestore.batch().apply {
                                delete(memberDoc.ref)

                                // Do the job of our trigger functions: remove this subject from
                                // user directly.
                                delete(firestore.document("users/$userId/subjects/$subjectId"))
                            }.commit()
                            supportFinishAfterTransition()
                        }
                        .build()

                md.getActionButton(DialogAction.POSITIVE).setTextColor(Palette.RED)
                md.show()
            }
            else -> return false
        }
        return true
    }

    override fun onStart() {
        super.onStart()
        manager.start()
    }

    override fun onStop() {
        manager.stop()
        super.onStop()
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        withAll({
            outState?.putParcelable(tag, value)
        }, memberDoc, subjectUser)
    }

    @SuppressLint("NewApi")
    private fun bind() {
        val state = manager.state.value
        val title: String
        when (state) {
            FragmentDocument.FireManager.ERROR -> {
                color = Palette.RED
                title = "Something went wrong"
            }
            else -> {
                val sub: ISubject? = subjectUser.value
                color = sub?.color ?: Palette.GREY
                title = sub?.name ?: if (state == FragmentDocument.FireManager.LOADING) {
                    UiHelper.TEXT_PLACEHOLDER
                } else "Subject not found"
            }
        }

        toolbarCollapsing.title = title
        refreshMenuItem.isVisible = state == FragmentDocument.FireManager.ERROR
        moderateMenuItem.isVisible = memberDoc.value?.role?.let { it >= Role.ADMIN } == true
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

    override fun onPageSelected(position: Int) {
        if (pagePosition == position) {
            return
        } else pagePosition = position

        val page = adapter.getPage(position) as? Page
        if (page != null) {
            val icon = page.hasFab()
            if (icon != 0) {
                fab.setImageResource(icon)
                fab.show()
            } else fab.hide()
        }
    }

    override fun onPageScrollStateChanged(state: Int) {}

}